cache/cache.go

```diff
@@ -0,0 +1,210 @@
+
+// ComponentBuilder creates CacheComponents for use in templates.
+//
+// See the package documentation for usage examples.
+type ComponentBuilder func(key string, opts ...Option) Component
+
+// NewCache creates a cache and returns a builder function
+// that can be used in templates. It accepts zero or more functional
+// options (WithTTL(), WithMaxMemory()).
+func NewCache(opts ...Option) ComponentBuilder {
+	base := Component{
+		ttl: defaultTTL,
+		lru: newLRU(defaultMem),
+	}
+
+	for _, opt := range opts {
+		opt(&base)
+	}
+	base.initialized = true
+
+	return func(key string, opts ...Option) Component {
+		dupe := base
+		dupe.key = key
+
+		for _, opt := range opts {
+			opt(&dupe)
+		}
+
+		return dupe
+	}
+}
+
+// WithTTL sets the default expiration duration for the cache,
+// or the expiration for an individual component. If the duration
+// is 0 then there is no expiration.
+func WithTTL(d time.Duration) Option {
+	return func(c *Component) {
+		if d == 0 {
+			d = 100 * 365 * 24 * time.Hour
+		}
+
+		c.ttl = d
+	}
+}
+
+// WithMaxMemory sets the maximum memory used for the cache. Note
+// that this will be ignored when set on individual components. If the
+// size is 0 then there is no memory limit.
+func WithMaxMemory(maxMem int) Option {
+	return func(c *Component) {
+		// This can't be changed after initialization
+		if c.initialized {
+			return
+		}
+
+		if maxMem == 0 {
+			maxMem = math.MaxInt
+		}
+
+		c.lru = newLRU(maxMem)
+	}
+}
+
+type Stats struct {
+	MaxMemory  int // maximum configured memory
+	UsedMemory int // memory used by cached items (including expired but not deleted items)
+	Items      int // cached item count (including expired but not deleted items)
+	Reads      int // total cache reads
+	Hits       int // total cache hits
+}
+
+// Stats returns basic cache statistics. These will be reset with Reset().
+func (c Component) Stats() Stats {
+	l := c.lru
+
+	return Stats{
+		MaxMemory:  l.maxMem,
+		UsedMemory: l.mem,
+		Items:      l.list.Len(),
+		Reads:      l.reads,
+		Hits:       l.hits,
+	}
+}
+
+// Remove removes/invalidates the cached data for associated with key, if it exists.
+func (c Component) Remove(key string) {
+	c.lru.deleteKey(key)
+}
+
+// Disable will turn off (or back on) caching. This also has the effect of wiping the cache.
+func (c *Component) Disable(disable bool) {
+	if disable {
+		c.lru.reset()
+	}
+
+	c.lru.disabled = disable
+}
+
+// Reset erases the cache and resets statistics.
+func (c *Component) Reset() {
+	c.lru.reset()
+}
+
+// Render will render child components, using cached data and caching results as needed.
+func (c Component) Render(ctx context.Context, w io.Writer) error {
+	cc, isCached := c.lru.get(c.key)
+	if isCached {
+		_, err := w.Write(cc)
+		return err
+	}
+
+	// Get children.
+	children := templ.GetChildren(ctx)
+	ctx = templ.ClearChildren(ctx)
+
+	if children == nil {
+		return nil
+	}
+
+	// Render children to a buffer.
+	var buf bytes.Buffer
```

Comment:
This is great for now, but I expect we'd want to use a buffer pool at some point, to reduce the amount of GC work required over time, since each new buffer is an allocation that needs to be cleared up.

Reply:
I've not used `sync.Pool` before. I thought this would be a straightforward change, but when I added it, my concurrency test started failing (not panicking, but rather the data rendered wasn't what was expected). This is the change: https://github.com/kalafut/templ-x/commit/f7cdafee96c7e7580ff154a8f7a526b024c6f2ad

Any idea what I'm doing wrong?